In [1]:
import json 
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd
import cv2
import os

In [2]:
path_to_ballet_2 ='C:\\Users\\user\\Desktop\\project\\1453169-20200914T174236Z-001\\TB_S'
path_9='C:\\Users\\user\\Desktop\\project\\1453169-20200914T174236Z-001\\SYN_B'

In [3]:
path_to_contemporary_2='C:\\Users\\user\\Desktop\\project\\1453169-20200914T174236Z-001\\TB_S_FB'

In [4]:
path_to_ballet = 'C:\\Users\\user\\Desktop\\project\\output\\ballet'
path_to_contemporary='C:\\Users\\user\\Desktop\\project\\output\\contemporary'
path_to_contemporary2='C:\\Users\\user\\Desktop\\project\\output\\contemporary21'
path_to_contemporary3='C:\\Users\\user\\Desktop\\project\\output\\contemporary22'

In [5]:
def read_keypoints(data):
    people_column=data['people']
    for d in people_column:
        keypoints=(d["pose_keypoints_2d"])
    return keypoints  


def remove_confidence_interval(data):

    j=2
    keypoints=read_keypoints(data)
    keypoints_woconfidence=keypoints.copy()
    while j<=len(keypoints_woconfidence):
        keypoints_woconfidence.pop(j)
        j+=2
    return keypoints_woconfidence

def create_y_coordicate(data):
    df=remove_confidence_interval(data)
    df=np.array(df)
    mask = np.ones(df.size, dtype=bool)
    mask[1::2] = 0
    points = df[mask]
    y=points.tolist()
    return y

def create_x_coordicate(data):
    df=remove_confidence_interval(data)
    df=np.array(df)
    mask = np.zeros(df.size, dtype=bool)
    mask[1::2] = 1
    points = df[mask]
    x=points.tolist()
    return x

def bodyparts(data):        
    
    # //     {0,  "Nose"}, 
    # //     {1,  "Neck"},
    # //     {2,  "RShoulder"},
    # //     {3,  "RElbow"},
    # //     {4,  "RWrist"},
    # //     {5,  "LShoulder"},
    # //     {6,  "LElbow"},
    # //     {7,  "LWrist"},
    # //     {8,  "MidHip"},
    # //     {9,  "RHip"},
    # //     {10, "RKnee"},
    # //     {11, "RAnkle"},
    # //     {12, "LHip"},
    # //     {13, "LKnee"},
    # //     {14, "LAnkle"},
    # //     {15, "REye"},
    # //     {16, "LEye"},
    # //     {17, "REar"},
    # //     {18, "LEar"},
    # //     {19, "LBigToe"},
    # //     {20, "LSmallToe"},
    # //     {21, "LHeel"},
    # //     {22, "RBigToe"},
    # //     {23, "RSmallToe"},
    # //     {24, "RHeel"},
    # //     {25, "Background"}    
    
    x=create_x_coordicate(data)
    y=create_y_coordicate(data)
    

    nose_to_neck=(x[0]-x[1]),(y[0]-y[1])
    
    right_eye_to_nose=(x[0]-x[15]),(y[0]-y[15])
    
    right_ear_to_eye=(x[15]-x[17]),(y[15]-y[17])
    left_eye_to_nose=(x[0]-x[16]),(y[0]-y[16])
    left_ear_to_eye=(x[16]-x[18]),(y[16]-y[18])
    
    torso= (x[1]-x[8]),(y[1]-y[8])
    
    
    right_shoulder=(x[1]-x[2]),(y[1]-y[2])
    left_shoulder=(x[1]-x[5]),(y[1]-y[5])
    
    right_upper_arm=(x[2]-x[3]),(y[2]-y[3])
    right_lower_arm=(x[3]-x[4]),(y[3]-y[4])
    left_upper_arm=(x[5]-x[6]),(y[5]-y[6])
    left_lower_arm=(x[6]-x[7]),(y[6]-y[7])
    
    right_hip=(x[8]-x[9]),(y[8]-y[9])
    right_upper_leg=(x[9]-x[10]),(y[9]-y[10])
    right_lower_leg=(x[10]-x[11]),(y[10]-y[11])
    left_hip=(x[8]-x[12]),(y[8]-y[12])
    left_upper_leg=(x[12]-x[13]),(y[12]-y[13])
    left_lower_leg=(x[13]-x[14]),(y[13]-y[14])
    
    left_foot=(x[14]-x[19]),(y[14]-y[19])
    left_toes=(x[19]-x[20]),(y[19]-y[20])
    left_ankle_to_heel=(x[14]-x[21]),(y[14]-y[21]) 
    right_foot=(x[11]-x[22]),(y[11]-y[22])
    right_toes=(x[22]-x[23]),(y[22]-y[23])
    right_ankle_to_heel=(x[11]-x[24]),(y[11]-y[24])
    
    body_vector=np.array([nose_to_neck, left_eye_to_nose ,left_ear_to_eye, right_eye_to_nose,right_ear_to_eye, torso,left_shoulder, right_shoulder,right_upper_arm ,right_lower_arm ,left_upper_arm ,left_lower_arm,left_hip ,left_upper_leg ,left_lower_leg ,right_hip ,right_upper_leg,right_lower_leg,right_foot ,right_toes ,right_ankle_to_heel ,left_foot ,left_toes ,left_ankle_to_heel])
    return body_vector


def similarity_bodyparts(data1,data2):
    #computes the similarity for every stick vector in the figures in one frame. 
    #I compute all the stick vectors per figure and then I do a pairwise comparison 
    #(also with cosine measure). 
    #I take the average of all similarities to come up with an overall similarity.
    bodyvector1=bodyparts(data1)
    bodyvector2=bodyparts(data2)
    cos_vector=np.array([])
    for i in range(len(bodyvector1)):
        a=cosine_similarity(bodyvector1[i].reshape(1,-1),bodyvector2[i].reshape(1,-1))
        #delete arrays with zeros
        if a != (np.array([0])):
            cos_vector=np.append(cos_vector,a )
        
    return round(np.mean(cos_vector),3) 

In [6]:
def create_df(path):
    newDF = pd.DataFrame(index=range(50))
    i=0
    
    
    for data in os.listdir(path):
        data=open(os.path.join(path,data),'r')
        data=json.load(data)
    
        def read_keypoints(data):
            people_column=data['people']
            for d in people_column:
                keypoints=(d["pose_keypoints_2d"])
            return keypoints  
    
    
        def remove_confidence_interval(data):
        
            j=2
            keypoints=read_keypoints(data)
            keypoints_woconfidence=keypoints.copy()
            while j<=len(keypoints_woconfidence):
                keypoints_woconfidence.pop(j)
                j+=2
            return keypoints_woconfidence
        
        
        bodyvector=remove_confidence_interval(data)
  
        new_bodyvector=pd.DataFrame(bodyvector)
        #print(new_bodyvector)
        
        
        #newDF['Frame: ',i]=new_bodyvector
        newDF[i]=new_bodyvector
        i+=1
        #print(bodyvector1)
        
    return newDF

In [7]:
create_df(path_9)

0        1        2        3        4        5        6        7   \
0   318.247  318.246  318.245  318.248  318.251  318.246  318.252  318.264   
1   121.251  121.250  121.260  121.254  121.267  121.269  121.268  121.265   
2   319.248  319.250  319.243  319.249  319.262  319.246  319.250  319.249   
3   170.264  170.266  170.259  170.257  170.254  170.245  170.242  170.243   
4   284.225  284.227  283.279  284.228  284.234  283.287  284.227  284.226   
5   170.256  170.259  170.247  170.238  170.235  170.234  170.234  170.229   
6   284.295  284.296  284.280  284.293  285.239  285.230  285.227  284.295   
7   231.256  231.255  230.273  230.278  229.289  230.231  230.239  229.265   
8   286.265  286.262  285.292  286.268  285.281  286.236  285.261  285.275   
9   261.250  261.260  261.253  262.269  261.249  261.259  261.240  260.272   
10  355.246  355.244  355.241  355.240  355.235  355.224  355.224  355.231   
11  170.234  170.237  170.240  170.236  170.228  170.227  169.281  170.221   
12  356.270  356.269  356.275  356.270  356.252  356.249  356.248  356.266   
13  232.245  232.245  231.279  232.256  233.227  233.248  233.271  233.236   
14  361.275  361.276  362.241  361.240  361.268  360.247  359.277  359.262   
15  263.249  263.246  262.275  262.250  262.247  264.272  264.256  263.273   
16  320.233  320.239  319.250  319.241  318.271  318.265  319.238  319.261   
17  265.244  265.238  265.243  265.272  266.267  266.258  266.247  266.228   
18  300.234  300.239  299.234  299.237  298.279  298.276  299.231  299.238   
19  264.260  264.254  264.269  265.232  265.271  265.264  265.265  265.236   
20  223.229  223.236  223.250  223.252  222.282  222.281  223.241  224.228   
21  242.237  242.232  242.234  242.228  241.261  241.244  240.271  240.242   
22  292.279  292.280  293.240  293.246  292.282  293.229  293.240  292.276   
23  292.228  292.227  292.239  292.251  292.227  292.237  292.237  291.282   
24  341.231  341.235  340.256  340.247  339.265  339.270  340.252  340.275   
25  264.264  264.257  264.265  265.247  266.233  265.272  265.254  265.245   
26  410.262  410.252  410.249  410.239  410.227  410.224  409.253  408.282   
27  237.250  237.239  237.233  236.293  237.231  237.231  236.287  236.264   
28  338.259  338.264  338.262  338.237  338.239  338.233  337.267  337.268   
29  293.252  293.247  293.251  293.272  293.273  293.263  293.273  293.259   
30  310.289  310.290  310.290  311.227  310.282  310.280  310.280  311.228   
31  114.273  114.273  114.276  114.269  114.276  114.277  114.274  114.272   
32  325.232  325.231  325.232  325.238  325.235  325.233  325.240  325.247   
33  114.248  114.247  114.252  114.247  114.247  114.251  114.247  114.241   
34  301.230  301.233  301.226  300.296  300.285  300.288  300.286  300.286   
35  122.221  122.223  121.282  121.280  122.225  122.221  122.225  122.222   
36  335.248  335.250  335.244  335.252  335.252  335.251  335.253  335.258   
37  119.259  119.255  119.253  119.247  119.241  119.246  119.241  119.236   
38  356.235  356.234  356.256  357.244  355.274  355.251  355.270  354.254   
39  306.244  306.247  306.241  305.279  306.234  306.236  306.253  306.269   
40  355.250  355.250  355.271  355.274  354.278  354.255  354.249  353.237   
41  302.262  302.265  302.257  302.238  302.239  302.242  302.253  302.283   
42  327.252  327.256  327.258  326.276  327.230  327.228  326.254  326.261   
43  297.264  297.256  297.261  298.238  297.273  297.262  297.275  297.262   
44  278.238  278.238  279.234  278.233  277.274  278.239  277.281  276.261   
45  304.242  304.244  304.235  304.244  303.269  303.273  303.278  303.276   
46  277.253  277.254  278.248  277.255  277.254  277.265  277.260  276.255   
47  302.258  302.260  302.264  302.258  302.233  302.233  302.245  301.282   
48  303.274  303.275  303.287  304.264  303.278  303.273  303.278  303.246   
49  295.250  295.250  295.250  295.273  295.245  295.245  295.237  294.284   

         8        9   ...     

In [8]:
#X1=create_df(path_to_ballet)
#X2=create_df(path_to_contemporary)

In [9]:
X1=create_df(path_9)
X2=create_df(path_to_contemporary_2)

In [10]:
x1_coordinates=X1[::2]
x2_coordinates=X2[::2]

In [11]:
x2_coordinates

0        1        2        3        4        5        6        7    \
0     0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000   
2   217.287  216.328  216.316  216.316  215.343  215.319  215.314  215.337   
4   238.308  237.343  237.324  237.308  236.342  236.326  236.316  236.321   
6   257.286  256.329  256.314  256.318  255.323  255.305  254.333  253.340   
8   242.307  242.306  242.311  242.294  240.332  239.306  238.346  238.300   
10  193.317  193.297  193.292  193.308  193.293  192.325  192.334  193.336   
12  149.326  148.352  149.290  148.313  148.294  148.295  147.299  147.313   
14  115.308  115.300  114.298  114.293  114.286  114.284  113.304  112.345   
16  218.303  218.304  217.315  217.315  216.328  216.294  215.332  215.341   
18  231.324  231.322  230.347  230.339  230.299  229.339  229.307  229.317   
20  219.293  219.352  221.290  221.300  221.345  222.298  222.308  222.330   
22  208.341  212.311  210.323  212.334  213.321  218.290  219.288  218.332   
24  204.321  204.327  203.308  203.311  202.324  201.339  201.326  201.326   
26  210.322  209.347  209.289  207.344  206.295  205.293  204.312  203.317   
28  218.347  219.293  219.321  219.309  219.321  219.293  219.279  218.330   
30    0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000   
32    0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000   
34  227.331  228.300  227.348  228.293  228.333  227.335  228.297  226.337   
36  203.293  202.341  203.298  203.336  203.336  203.339  203.302  203.314   
38  199.308  200.292  199.334  200.322  200.298  199.308  199.330  199.337   
40  202.346  203.304  203.301  203.302  203.307  203.288  203.296  203.295   
42  222.325  222.332  222.352  222.334  223.322  223.304  222.342  222.316   
44  200.308  200.290  201.316  199.333  201.321  202.317  202.322  202.293   
46  203.332  203.308  204.349  203.324  204.343  205.297  205.300  204.320   
48  221.334  218.321  209.330  212.337  213.322  225.307  226.297  226.290   

        8        9    ...      91       92       93       94       95   \
0     0.000    0.000  ...    0.000    0.000    0.000    0.000    0.000   
2   215.331  215.331  ...  253.353  254.305  254.307  254.300  254.310   
4   236.325  236.314  ...  274.288  274.309  274.302  274.302  274.324   
6   252.353  252.307  ...  293.297  292.330  292.338  293.296  293.311   
8   238.320  236.314  ...  307.314  307.342  308.329  308.338  308.328   
10  193.328  193.339  ...  234.317  234.319  234.320  234.329  234.339   
12  147.329  147.334  ...  209.314  209.323  208.348  209.300  209.325   
14  111.337  110.346  ...  193.332  194.285  193.309  193.309  193.332   
16  214.335  214.307  ...  254.327  254.348  254.336  254.340  255.296   
18  228.337  228.324  ...  267.332  267.345  267.329  267.335  267.341   
20  222.349  222.340  ...  270.296  268.356  267.355  266.323  265.308   
22  219.285  218.323  ...  251.274  250.340  251.286  251.295  250.342   
24  200.322  199.336  ...  242.297  242.316  242.310  242.315  242.339   
26  202.334  201.324  ...  238.335  239.346  241.290  242.302  243.315   
28  218.291  218.330  ...  261.306  261.309  260.352  260.353  260.348   
30    0.000    0.000  ...    0.000    0.000    0.000    0.000    0.000   
32    0.000    0.000  ...    0.000    0.000    0.000    0.000    0.000   
34  227.319  226.308  ...  264.315  264.302  263.324  263.326  263.324   
36  203.330  203.308  ...  243.316  243.349  244.285  243.300  243.318   
38  199.306  199.344  ...  245.351  245.346  244.345  244.323  244.334   
40  203.289  203.325  ...  248.316  248.303  246.332  246.327  246.323   
42  221.349  222.325  ...  265.313  265.311  264.344  264.344  264.324   
44  201.334  201.324  ...  239.323  239.344  239.320  239.304  239.319   
46  204.292  204.302  ...  242.335  243.321  243.308  242.336  243.317   
48  226.283  226.275  ...  250.295  249.335  250.304  250.320  250.285   

        96       97       98       99       100  
0   

In [12]:
x1_coordinates

0        1        2        3        4        5        6        7   \
0   318.247  318.246  318.245  318.248  318.251  318.246  318.252  318.264   
2   319.248  319.250  319.243  319.249  319.262  319.246  319.250  319.249   
4   284.225  284.227  283.279  284.228  284.234  283.287  284.227  284.226   
6   284.295  284.296  284.280  284.293  285.239  285.230  285.227  284.295   
8   286.265  286.262  285.292  286.268  285.281  286.236  285.261  285.275   
10  355.246  355.244  355.241  355.240  355.235  355.224  355.224  355.231   
12  356.270  356.269  356.275  356.270  356.252  356.249  356.248  356.266   
14  361.275  361.276  362.241  361.240  361.268  360.247  359.277  359.262   
16  320.233  320.239  319.250  319.241  318.271  318.265  319.238  319.261   
18  300.234  300.239  299.234  299.237  298.279  298.276  299.231  299.238   
20  223.229  223.236  223.250  223.252  222.282  222.281  223.241  224.228   
22  292.279  292.280  293.240  293.246  292.282  293.229  293.240  292.276   
24  341.231  341.235  340.256  340.247  339.265  339.270  340.252  340.275   
26  410.262  410.252  410.249  410.239  410.227  410.224  409.253  408.282   
28  338.259  338.264  338.262  338.237  338.239  338.233  337.267  337.268   
30  310.289  310.290  310.290  311.227  310.282  310.280  310.280  311.228   
32  325.232  325.231  325.232  325.238  325.235  325.233  325.240  325.247   
34  301.230  301.233  301.226  300.296  300.285  300.288  300.286  300.286   
36  335.248  335.250  335.244  335.252  335.252  335.251  335.253  335.258   
38  356.235  356.234  356.256  357.244  355.274  355.251  355.270  354.254   
40  355.250  355.250  355.271  355.274  354.278  354.255  354.249  353.237   
42  327.252  327.256  327.258  326.276  327.230  327.228  326.254  326.261   
44  278.238  278.238  279.234  278.233  277.274  278.239  277.281  276.261   
46  277.253  277.254  278.248  277.255  277.254  277.265  277.260  276.255   
48  303.274  303.275  303.287  304.264  303.278  303.273  303.278  303.246   

         8        9   ...       29       30       31       32       33  \
0   318.272  318.282  ...  320.277  321.221  320.272  320.273  320.273   
2   319.257  319.254  ...  321.243  321.245  321.246  321.233  321.234   
4   284.236  284.225  ...  284.273  284.274  284.272  284.255  284.262   
6   285.225  284.290  ...  285.266  285.256  285.261  285.266  285.267   
8   285.282  285.281  ...  285.238  285.236  284.269  284.278  284.289   
10  355.238  355.246  ...  357.252  357.252  357.253  357.250  357.256   
12  356.286  356.295  ...  357.252  357.242  357.245  357.242  357.249   
14  359.270  361.266  ...  367.248  367.241  367.235  366.289  367.237   
16  319.230  318.231  ...  319.271  319.280  319.279  319.265  319.274   
18  299.233  297.269  ...  299.276  299.276  299.278  299.270  299.272   
20  225.264  227.230  ...  295.268  294.260  295.241  295.255  294.277   
22  293.232  292.260  ...    0.000    0.000    0.000  294.262    0.000   
24  340.237  339.255  ...  340.257  340.269  340.268  340.262  340.275   
26  407.273  405.280  ...  333.268  331.264  351.243  345.250  334.252   
28  338.228  337.260  ...    0.000    0.000  333.245  331.262    0.000   
30  311.234  311.249  ...  313.242  313.248  313.244  313.240  313.230   
32  325.254  325.265  ...  327.260  327.262  327.256  327.256  327.250   
34  300.287  301.227  ...  303.235  303.233  303.234  302.287  302.282   
36  335.256  335.254  ...  338.246  338.246  338.235  338.233  338.226   
38  354.275  357.238  ...    0.000    0.000  325.260  325.265    0.000   
40  353.253  355.241  ...    0.000    0.000  331.271  332.252    0.000   
42  326.268  326.261  ...    0.000    0.000  326.258  326.252    0.000   
44  275.268  274.259  ...    0.000    0.000    0.000  291.271    0.000   
46  275.270  274.271  ...    0.000    0.000    0.000    0.000    0.000   
48  303.257  302.293  ...    0.000    0.000    0.000  296.276    0.000   

         34       35       36       37       38  
0   3

In [13]:
y1_coordinates=X1[1::2]
y2_coordinates=X2[1::2]

In [15]:
from tslearn.metrics import dtw, dtw_path
dtw(X1,X2)

5372.339172613242

In [16]:
path=dtw_path(X1,X2)[0]

In [17]:
L=path

In [14]:
L=[(0, 0),
 (1, 1),
 (2, 1),
 (3, 2),
 (4, 3),
 (5, 4),
 (6, 5),
 (7, 6),
 (8, 7),
 (9, 8),
 (10, 9),
 (11, 10),
 (12, 11),
 (13, 12),
 (14, 13),
 (15, 14),
 (16, 14),
 (17, 15),
 (18, 16),
 (19, 17),
 (20, 18),
 (21, 19),
 (22, 20),
 (23, 20),
 (24, 20),
 (25, 20),
 (26, 20),
 (27, 21),
 (28, 22),
 (29, 23),
 (30, 24),
 (30, 25),
 (31, 26),
 (32, 26),
 (33, 27),
 (34, 28),
 (35, 29),
 (36, 30),
 (37, 31),
 (38, 32),
 (39, 33),
 (40, 33),
 (41, 33),
 (42, 34),
 (43, 35),
 (44, 36),
 (45, 37),
 (46, 37),
 (47, 38),
 (48, 39),
 (49, 40),
 (50, 41),
 (51, 42),
 (52, 43),
 (53, 44),
 (54, 45),
 (55, 46),
 (56, 46),
 (57, 47),
 (58, 47),
 (59, 47),
 (60, 47),
 (61, 48),
 (62, 48),
 (63, 48),
 (64, 49),
 (65, 50),
 (66, 51),
 (67, 51),
 (68, 51),
 (69, 52),
 (70, 53),
 (71, 54),
 (72, 55),
 (73, 56),
 (74, 57),
 (75, 58),
 (76, 59),
 (77, 60),
 (78, 61),
 (79, 62),
 (80, 63),
 (81, 64),
 (82, 65),
 (83, 66),
 (84, 67),
 (85, 68),
 (86, 69),
 (87, 70),
 (88, 71),
 (89, 72),
 (90, 73),
 (91, 74),
 (92, 75),
 (93, 76),
 (94, 77),
 (95, 78),
 (96, 79),
 (97, 80),
 (97, 81),
 (97, 82),
 (97, 83),
 (97, 84),
 (97, 85),
 (97, 86),
 (97, 87),
 (97, 88),
 (97, 89),
 (97, 90),
 (97, 91),
 (97, 92),
 (97, 93),
 (97, 94),
 (97, 95),
 (97, 96),
 (97, 97),
 (97, 98),
 (97, 99),
 (98, 100)]

# Similarity for alligned frames

συγκρινεις Χ1,Χ2 που εχουν και χ και y αλλα δε σε νοιαζει γιατι το κανει pairwise ouτως ή αλλως.
το μηκος δεν ειναι το ιδιο...πολλα frames αντιστοιχουν σε ενα. αρα ειναι πανω απο 122 οι αντιστοιχισεις

In [62]:
len(X2)

50

In [19]:
similar_num=0
cos_vector=np.array([])
for i,j in L:
    a=cosine_similarity(X1[i].values.reshape(1,-1),X2[j].values.reshape(1,-1))
    #delete arrays with zeros
    if a != (np.array([0])):
        cos_vector=np.append(cos_vector,a )
        
print(len(cos_vector))
similar_num=round(np.mean(cos_vector),3) 

ValueError: Incompatible dimension for X and Y matrices: X.shape[1] == 39 while Y.shape[1] == 101

In [20]:
similar_num

0

In [64]:
cos_vector

array([0.95018651, 0.95038459, 0.95102381, 0.95144317, 0.95189914,
       0.95246311, 0.93925203, 0.94132997, 0.94524226, 0.94585528,
       0.95607098, 0.95654477, 0.95125209, 0.95855404, 0.9582806 ,
       0.93630951, 0.16001209, 0.97131848, 0.99498804, 0.99551869,
       0.97225255, 0.95869709, 0.93862097, 0.93834303, 0.97293013,
       0.99688682, 0.99699676, 0.99767647, 0.97563845, 0.99795001,
       0.9632737 , 0.96337769, 0.99818138, 0.92738162, 0.99720206,
       0.96693974, 0.99657439, 0.99557954, 0.99574029, 0.99557205,
       0.98349153, 0.99619727, 0.99620222, 0.94871927, 0.20135696,
       0.92129111, 0.92423723, 0.91944029, 0.92706479, 0.90169633,
       0.97142297, 0.95643406, 0.98413595, 0.98404554, 0.98430693,
       0.99854138, 0.98482274, 0.98509336, 0.9991084 , 0.99915028,
       0.99924378, 0.99933283, 0.99919919, 0.99923655, 0.99889602,
       0.98606057, 0.98569257, 0.9855928 , 0.98543407, 0.98543977,
       0.97180094, 0.97134393, 0.9707071 , 0.97041687, 0.98288

In [65]:
np.max(cos_vector)

0.9993328340928246

In [66]:
np.argmax(cos_vector)

61

61th frame of video1.

Most important frames that contribute to similarity??

In [67]:
cos_vector.argsort()[-10:][::-1]

array([61, 60, 63, 62, 59, 58, 64, 55, 32, 29], dtype=int64)

How to determine the most important angle in terms of similarity?


In [68]:
import numpy as np
from sklearn.decomposition import PCA

pca = PCA(n_components=10)
pca.fit(X1)
PCA(n_components=2)
print(pca.explained_variance_ratio_)

print(pca.singular_values_)


[0.75374833 0.10020958 0.05669691 0.02209361 0.0170967  0.0154332
 0.0100103  0.00724448 0.00494226 0.00337795]
[4976.44101059 1814.51394072 1364.85135892  851.9989144   749.48313656
  712.08800858  573.49446132  487.87576067  402.96592201  333.14423233]
